map.go

```diff
@@ -0,0 +1,141 @@
+package surge
+
+import (
+	"reflect"
+	"sort"
+	"unsafe"
+)
+
+func sizeHintReflectedMap(v reflect.Value) int {
+	sizeHint := 2
+	iter := v.MapRange()
+	for iter.Next() {
+		sizeHint += sizeHintReflected(iter.Key())
+		sizeHint += sizeHintReflected(iter.Value())
+	}
+	return sizeHint
+}
+
+func marshalReflectedMap(v reflect.Value, buf []byte, rem int) ([]byte, int, error) {
+	buf, rem, err := MarshalU16(uint16(v.Len()), buf, rem)
+	if err != nil {
+		return buf, rem, err
+	}
+
+	// Define a local key/value type that can be used to sort the key/value
+	// pairs. It is worth noting, the key is explicitly defined to be bytes, so
+	// we will need to marshal all keys and then convert those bytes to strings.
+	// This guarantees that, regardless of the key type, we can sort the keys
+	// deterministically.
+	type KeyValue struct {
+		keyData []byte
+		key     reflect.Value
+	}
+
+	// Allocate a slice for storing the key/value pairs. This is needed so that
+	// we can then sort the slice, ensuring that the key/value ordering is
+	// deterministic.
+	n := v.Len() * int(unsafe.Sizeof(KeyValue{}))
+	if n < 0 {
```

Comment:
@ross-pure it could be, but `rem < n` is checked on the next line, which catches the case that we care about (where overflow causes the length the be unacceptably large based on `rem`)

Reply:
@loongy But I think that you could have some large `v.Len()` that causes `n` to overflow to a *small* positive number. However, I've realised that since this is in the marshal function it is less relevant, but then the user of the function would only be sabotaging themself, and it would require probably a very large map.

Reply:
@ross-pure pushed a fix: `if n < 0 || n < v.Len() { .. }`

Reply:
@loongy The new condition makes the first part of the or irrelevant I think (if `v.Len()` is always non-negative), but I think this check is not crucial as I said anyway so I'm going to approve.